File: backend/app/schemas/food.py
from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel


class FoodAnalysisResponse(BaseModel):
    name: str
    calories: int
    protein: float | None = None
    fat: float | None = None
    carbs: float | None = None
    grams: float | None = None
    confidence: float | None = None


class FoodEntryCreate(BaseModel):
    name: str
    calories: int
    protein: float | None = None
    fat: float | None = None
    carbs: float | None = None
    grams: float | None = None
    photo_url: str | None = None


class IngredientCreate(BaseModel):
    name: str
    calories: int
    protein: float | None = None
    fat: float | None = None
    carbs: float | None = None
    grams: float | None = None


class FoodEntryWithIngredientsCreate(BaseModel):
    name: str
    calories: int
    protein: float | None = None
    fat: float | None = None
    carbs: float | None = None
    grams: float | None = None
    photo_url: str | None = None
    ingredients: list[IngredientCreate] = []


class IngredientResponse(BaseModel):
    id: UUID
    name: str
    calories: int
    protein: float | None = None
    fat: float | None = None
    carbs: float | None = None
    grams: float | None = None

    class Config:
        from_attributes = True


class FoodEntryResponse(BaseModel):
    id: UUID
    user_id: UUID
    parent_id: UUID | None = None
    name: str
    calories: int
    protein: float | None = None
    fat: float | None = None
    carbs: float | None = None
    grams: float | None = None
    photo_url: str | None = None
    created_at: datetime
    updated_at: datetime
    ingredients: list[IngredientResponse] = []

    class Config:
        from_attributes = True


class DailyStatsResponse(BaseModel):
    date: date
    total_calories: int
    total_protein: float
    total_fat: float
    total_carbs: float
    remaining_calories: int | None = None
    calorie_goal: int | None = None
    entries_count: int


class RecommendationResponse(BaseModel):
    recommendations: list[str]
    insights: str
    protein_status: str
    balance_score: int


class FoodEntriesListResponse(BaseModel):
    entries: list[FoodEntryResponse]
    total: int
    page: int
    page_size: int


class FoodItemAnalysis(BaseModel):
    """Single ingredient/product detected in the dish"""

    name: str
    calories: int
    protein: float | None = None
    fat: float | None = None
    carbs: float | None = None
    grams: float | None = None


class DishAnalysisResponse(BaseModel):
    """Full dish analysis with ingredient breakdown"""

    dish_name: str
    total_calories: int
    total_grams: float
    total_protein: float
    total_fat: float
    total_carbs: float
    items: list[FoodItemAnalysis]
    confidence: float | None = None


class AnalyzePhotoDetailedResponse(DishAnalysisResponse):
    """API response with photo URL"""

    photo_url: str

File: frontend/src/types/food.ts
export interface FoodAnalysisResponse {
  name: string;
  calories: number;
  protein: number | null;
  fat: number | null;
  carbs: number | null;
  grams: number | null;
  confidence: number | null;
  photo_url: string;
}

export interface FoodItemAnalysis {
  name: string;
  calories: number;
  protein: number | null;
  fat: number | null;
  carbs: number | null;
  grams: number | null;
}

export interface DishAnalysisResponse {
  dish_name: string;
  total_calories: number;
  total_grams: number;
  total_protein: number;
  total_fat: number;
  total_carbs: number;
  items: FoodItemAnalysis[];
  confidence: number | null;
  photo_url: string;
}

export interface FoodEntryCreate {
  name: string;
  calories: number;
  protein?: number | null;
  fat?: number | null;
  carbs?: number | null;
  grams?: number | null;
  photo_url?: string | null;
}

export interface IngredientCreate {
  name: string;
  calories: number;
  protein?: number | null;
  fat?: number | null;
  carbs?: number | null;
  grams?: number | null;
}

export interface FoodEntryWithIngredientsCreate {
  name: string;
  calories: number;
  protein?: number | null;
  fat?: number | null;
  carbs?: number | null;
  grams?: number | null;
  photo_url?: string | null;
  ingredients: IngredientCreate[];
}

export interface Ingredient {
  id: string;
  name: string;
  calories: number;
  protein: number | null;
  fat: number | null;
  carbs: number | null;
  grams: number | null;
}

export interface FoodEntry {
  id: string;
  user_id: string;
  parent_id: string | null;
  name: string;
  calories: number;
  protein: number | null;
  fat: number | null;
  carbs: number | null;
  grams: number | null;
  photo_url: string | null;
  created_at: string;
  updated_at: string;
  ingredients: Ingredient[];
}

export interface FoodEntriesListResponse {
  entries: FoodEntry[];
  total: number;
  page: number;
  page_size: number;
}

export interface DailyStats {
  date: string;
  total_calories: number;
  total_protein: number;
  total_fat: number;
  total_carbs: number;
  remaining_calories: number | null;
  calorie_goal: number | null;
  entries_count: number;
}

export interface RecommendationResponse {
  recommendations: string[];
  insights: string;
  protein_status: string;
  balance_score: number;
}

export interface NutritionSearchResult {
  name: string;
  calories: number;
  protein_g: number;
  fat_total_g: number;
  carbohydrates_total_g: number;
  serving_size_g: number;
}

export interface NutritionSearchResponse {
  query: string;
  results: NutritionSearchResult[];
}
